pkg/pki/ca/ca.go

```diff
@@ -56,24 +70,57 @@ type IstioCA struct {
 }
 
 // NewSelfSignedIstioCA returns a new IstioCA instance using self-signed certificate.
-func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioCA, error) {
-	now := time.Now()
-	options := CertOptions{
-		NotBefore:    now,
-		NotAfter:     now.Add(caCertTTL),
-		Org:          org,
-		IsCA:         true,
-		IsSelfSigned: true,
-		RSAKeySize:   caKeySize,
-	}
-	pemCert, pemKey := GenCert(options)
+func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string, namespace string,
+	core corev1.SecretsGetter) (*IstioCA, error) {
 
+	// For the first time the CA is up, it generates a self-signed key/cert pair and write it to
+	// cASecret. For subsequent restart, CA will reads key/cert from cASecret.
+	caSecret, err := core.Secrets(namespace).Get(cASecret, metav1.GetOptions{})
 	opts := &IstioCAOptions{
-		CertTTL:          certTTL,
-		SigningCertBytes: pemCert,
-		SigningKeyBytes:  pemKey,
-		RootCertBytes:    pemCert,
+		CertTTL: certTTL,
+	}
+	if err != nil {
+		glog.Infof("Failed to get secret (error: %s), will create one", err)
+
+		now := time.Now()
+		options := CertOptions{
+			NotBefore:    now,
+			NotAfter:     now.Add(caCertTTL),
+			Org:          org,
+			IsCA:         true,
+			IsSelfSigned: true,
+			RSAKeySize:   caKeySize,
+		}
+		pemCert, pemKey := GenCert(options)
+
+		opts.SigningCertBytes = pemCert
+		opts.SigningKeyBytes = pemKey
+		opts.RootCertBytes = pemCert
+
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				cACertID:       pemCert,
+				cAPrivateKeyID: pemKey,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cASecret,
+				Namespace: namespace,
+			},
+			Type: istioSecretType,
+		}
+		_, err := core.Secrets(namespace).Create(secret)
+		if err != nil {
+			return nil, err
```

Comment:
IMHO - log an error ("can't save the cert") - but keep running. Even if the CA restart, we will eventually recover - however if CA fails to start 
pretty much everything will stop working.

Reply:
Good point, done.